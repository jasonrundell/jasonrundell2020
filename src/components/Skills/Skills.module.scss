.list {
  display: flex;
  flex-direction: column;
  margin-top: 0;
  margin-bottom: $sizeNormal;
  list-style: none;
  padding-left: 0;
}

.title {
  margin-top: 0;
  margin-bottom: $sizeSmaller;
  font-weight: 700;
}

.item {
  display: flex;
  flex-direction: row;
  width: 100%;
  margin-bottom: $sizeSmallest;
  .item-text {
    flex: 1;
    font-size: $sizeSmall;
  }
}
