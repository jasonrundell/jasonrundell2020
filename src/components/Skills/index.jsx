import React from 'react'
import PropTypes from 'prop-types'
import { StaticQuery, graphql } from 'gatsby'

import Layout from '../Layout'
import Badge from '../Badge'
import Lists from '../Lists'
import onlyUnique from '../../utils/unique'

// CSS
import styles from './Skills.module.scss'

const { Box, Grid, Row } = Layout
const { BadgeList } = Lists

const Skills = ({ data }) => {
  const edges = data.allContentfulSkills.edges
  const allSkills = [...edges]
  let categories = []
  let skills = []

  allSkills.forEach(category => {
    categories.push(category.node.category)
  })

  // I only want the unique categories
  const uniqueCategories = categories.filter(onlyUnique)

  allSkills.forEach(skill => {
    skills.push(skill.node)
  })

  const legendList = [
    {
      id: 'expert',
      icon: 'E',
      label: 'Expert',
    },
    {
      id: 'advanced',
      icon: 'A',
      label: 'Advanced',
    },
    {
      id: 'proficient',
      icon: 'P',
      label: 'Proficient',
    },
    {
      id: 'intermediate',
      icon: 'I',
      label: 'Intermediate',
    },
  ]

  return (
    <>
      <Row>
        <BadgeList list={legendList} isHorizontal isInverse />
      </Row>
      <Row>
        <Grid
          columnCount={1}
          mediumColumnCount={1}
          largeColumnCount={3}
          breakInside="avoid"
        >
          {uniqueCategories.map((parentCategory, index) => {
            return (
              <div key={index} className={styles.list}>
                <h3 className={styles.title}>{parentCategory}</h3>
                <ul className={styles.list}>
                  {skills.map(item => {
                    const { id, name, category, expertiseLevel } = item

                    if (parentCategory === category) {
                      return (
                        <li key={id} className={styles.item}>
                          <Box isTight>
                            <Badge icon={expertiseLevel} isInverse />
                          </Box>
                          <Box isTight>
                            <span className={styles.itemText}>{name}</span>
                          </Box>
                        </li>
                      )
                    } else {
                      return null
                    }
                  })}
                </ul>
              </div>
            )
          })}
        </Grid>
      </Row>
    </>
  )
}

export default props => (
  <StaticQuery
    query={graphql`
      {
        allContentfulSkills(sort: { fields: category }) {
          edges {
            node {
              id
              category
              name
              expertiseLevel
            }
          }
        }
      }
    `}
    render={data => <Skills data={data} {...props} />}
  />
)
Skills.propTypes = {
  data: PropTypes.shape({
    allContentfulSkills: PropTypes.shape({
      edges: PropTypes.arrayOf(PropTypes.object).isRequired,
    }).isRequired,
  }).isRequired,
}
