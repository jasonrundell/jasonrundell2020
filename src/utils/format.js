/**
 *  Letter case conversion functions
 *  [Letter case](https://en.wikipedia.org/wiki/Letter_case#Special_case_styles)
 *
 *  Converts only letter case-separated or hyphen-delimited strings
 */

/**
 *  Converts hyphen-delimited `string` to camel case
 *
 *  @example
 *  'foo-bar' => 'fooBar'
 *  'foo--bar' => 'fooBar'
 *
 *  @param {string} string
 */
function toCamelCase(string) {
  return string.replace(/(-+)([a-z])/g, ($0, $1, $2) => $2.toUpperCase())
}

/**
 *  Converts hyphen-delimited `string` to pascal case (aka upper camel case)
 *
 *  @example
 *  'foo-bar' => 'FooBar'
 *  'foo--bar' => 'FooBar'
 *
 *  @param {string} string
 */
function toPascalCase(string) {
  return toCamelCase(string.replace(/^[a-z]/, $0 => $0.toUpperCase()))
}

export { toCamelCase, toPascalCase }
