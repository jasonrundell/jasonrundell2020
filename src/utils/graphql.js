export function edgesToObject(edgesData) {
  const nodes = edgesData.map(item => {
    return item.node
  })
  const finalObject = nodes.map(item => {
    return item
  })

  return finalObject
}
