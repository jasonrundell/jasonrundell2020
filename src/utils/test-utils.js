export function catchWarnings() {
  const originalConsoleError = global.console.error
  beforeEach(() => {
    /**
     * I Want to know if values passed to props fail the type check
     * */
    global.console.error = (...args) => {
      const propTypeFailures = [/Warning:/] // e.g. [/Invalid prop/, /Failed prop type/]

      if (propTypeFailures.some(p => p.test(args[0]))) {
        throw new Error(args[0])
      }

      originalConsoleError(...args)
    }
  })
}

/**
 * Mix an match help from:
 * https://daveceddia.com/testing-for-proptypes-errors/
 * https://www.jackfranklin.co.uk/blog/failing-tests-on-react-proptypes/
 */
