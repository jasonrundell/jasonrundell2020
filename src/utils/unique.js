const onlyUnique = (value, index, self) => {
  return self.indexOf(value) === index
}

const generatePassword = ({
  passwordLength = 16,
  charset = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*()_-=~`',
}) => {
  let retVal = ''
  for (var i = 0, n = charset.length; i < passwordLength; ++i) {
    retVal += charset.charAt(Math.floor(Math.random() * n))
  }
  return retVal
}

export { onlyUnique, generatePassword }
