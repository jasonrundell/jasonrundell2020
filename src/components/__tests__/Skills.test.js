import React from 'react'
import { render, unmountComponentAtNode } from 'react-dom'
import { act } from 'react-dom/test-utils'
import renderer from 'react-test-renderer'
import { catchWarnings } from '../../utils/test-utils'

import Skills from '../Skills'

catchWarnings()

let container = null
beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  unmountComponentAtNode(container)
  container.remove()
  container = null
})

const testData = [
  {
    id: 'eb882bd6-c391-5613-9795-2dd6cd134e54',
    category: 'APIs',
    name: 'Facebook',
  },
  {
    id: 'f4f7b181-cc7d-57ce-9d59-15a9916fbec1',
    category: 'APIs',
    name: 'Slack',
  },
  {
    id: 'd4b52d23-aa82-5539-badb-48382003d372',
    category: 'APIs',
    name: 'Google Places',
  },
]

describe('Skills component', () => {
  it('Skills renders correctly', () => {
    const tree = renderer.create(<Skills skills={testData} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Skills has required skills prop', () => {
    act(() => {
      render(<Skills skills={testData} />, container)
    })
  })
})
