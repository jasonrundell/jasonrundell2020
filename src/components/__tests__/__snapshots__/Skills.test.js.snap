// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skills component Skills renders correctly 1`] = `
Array [
  <div
    className="root start"
  >
    <ul
      className="root horizontal"
    >
      <li
        className="listItem"
      >
        <div
          className="root tight"
        >
          <div
            className="root inverse"
          >
            <div
              className="root tight"
            >
              E
            </div>
          </div>
        </div>
        <div
          className="root tight"
        >
          <span>
            Expert
          </span>
        </div>
      </li>
      <li
        className="listItem"
      >
        <div
          className="root tight"
        >
          <div
            className="root inverse"
          >
            <div
              className="root tight"
            >
              A
            </div>
          </div>
        </div>
        <div
          className="root tight"
        >
          <span>
            Advanced
          </span>
        </div>
      </li>
      <li
        className="listItem"
      >
        <div
          className="root tight"
        >
          <div
            className="root inverse"
          >
            <div
              className="root tight"
            >
              P
            </div>
          </div>
        </div>
        <div
          className="root tight"
        >
          <span>
            Proficient
          </span>
        </div>
      </li>
      <li
        className="listItem"
      >
        <div
          className="root tight"
        >
          <div
            className="root inverse"
          >
            <div
              className="root tight"
            >
              I
            </div>
          </div>
        </div>
        <div
          className="root tight"
        >
          <span>
            Intermediate
          </span>
        </div>
      </li>
    </ul>
  </div>,
  <div
    className="root start"
  >
    <div
      className="root col1 colMedium2 colLarge3 breakInside__avoid"
    >
      <div
        className="list"
      >
        <h2
          className="h2"
        >
          APIs
        </h2>
        <ul
          className="list"
        >
          <li
            className="item"
          >
            <div
              className="root tight"
            >
              <div
                className="root inverse"
              >
                <div
                  className="root tight"
                >
                  P
                </div>
              </div>
            </div>
            <div
              className="root tight"
            >
              <span
                className="itemText"
              >
                Facebook
              </span>
            </div>
          </li>
          <li
            className="item"
          >
            <div
              className="root tight"
            >
              <div
                className="root inverse"
              >
                <div
                  className="root tight"
                >
                  I
                </div>
              </div>
            </div>
            <div
              className="root tight"
            >
              <span
                className="itemText"
              >
                Slack
              </span>
            </div>
          </li>
          <li
            className="item"
          >
            <div
              className="root tight"
            >
              <div
                className="root inverse"
              >
                <div
                  className="root tight"
                >
                  P
                </div>
              </div>
            </div>
            <div
              className="root tight"
            >
              <span
                className="itemText"
              >
                Google Places
              </span>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>,
]
`;
